harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -788,11 +1015,38 @@ def _load(self, load_path: pathlib.Path) -> None:
                     "callback will be initialized from scratch"
                 )
 
-        # Load workload sequencer state.
-        wlsq_path = load_path.joinpath("workload_sequencer.pkl")
-        if self.wlsq is not None and wlsq_path.exists():
-            with wlsq_path.open("rb") as f:
-                self.wlsq.load_state(pickle.load(f))
+        save_path = load_path.joinpath("trial_state.pkl")
+        if save_path.exists():
+            with save_path.open("rb") as f:
+                self.load_state(pickle.load(f))
+        else:
+            # Support legacy save states
+            wlsq_path = load_path.joinpath("workload_sequencer.pkl")
+            if wlsq_path.exists():
+                with wlsq_path.open("rb") as f:
+                    self.load_state(pickle.load(f))
+
+    def load_state(self, state: Any) -> None:
+        # Load our state from the checkpoint if we are continuing training after a pause or restart.
+        # If the trial_id doesn't match our current trial id, we're continuing training a previous
+        # trial and the state in the checkpoint should be discarded.
+        if state.get("trial_id") != self._trial_id:
+            return
+
+        self.state = TrialState(**state)
+        self.last_checkpoint = self.state.last_ckpt
+        self.last_validation = self.state.last_val
+        self.steps_completed = self.state.steps_completed
+        self.epochs_trained = self.state.epochs_trained
+        self.batches_trained = self.state.batches_trained
+
+        # Detect the case where the final validation we made was against this exact checkpoint.  In
+        # that case, the master will know about the validation, but it would not appear in the
+        # checkpoint state.  If the validation was before the last checkpoint, the checkpoint state
+        # is already correct, while any validations after the last checkpoint aren't valid anymore
+        # and can be safely ignored.
+        if self.steps_completed == self.val_from_previous_run:
+            self.last_validation = self.steps_completed
```

Comment:
question: does this training loop validate after checkpointing?  If not, we can get rid of this ugly complexity.